### PR TITLE
Have "Handle Fetch" return error when request's streaming body is unusable

### PR DESCRIPTION
**Pull request by yutakahirano**
***
<a href="https://pr-preview.s3.amazonaws.com/yutakahirano/ServiceWorker/pull/1563.html" title="Last updated on Feb 18, 2021, 4:09 AM UTC (40e9e10)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/ServiceWorker/1563/044cc99...yutakahirano:40e9e10.html" title="Last updated on Feb 18, 2021, 4:09 AM UTC (40e9e10)">Diff</a>

### CODE REVIEW

**Comment by yutakahirano** [comment]
@annevk @yoichio

This is related to whatwg/fetch#1144.

File: docs/index.bs
Comment (annevk) [inline docs/index.bs:2952]:
```suggestion
      :: |response|, null or a [=/response=]
```

File: docs/index.bs
Comment (annevk) [inline docs/index.bs:2953]:
```suggestion
      :: |requestBody|, null or a [=/body=]
```

File: docs/index.bs
Comment (annevk) [inline docs/index.bs:3025]: Shouldn't this happen much earlier? We need to essentially transfer the request body back. If we just return it there are no side effects for the service worker side and there need to be when it's backed by a stream, right?

Comment (yutakahirano) [reply docs/index.bs:3025]: Which point do you think is the best?

Do you mean [transferring stream](https://github.com/whatwg/streams/issues/276) by "transfer"? I think what's needed is a bit different from that transferring as 

 - We don't keep chunk boundary.
 - Chunks other than Uint8Arrays are rejected.

I think we need to define a means to create such a Body and/or ReadableStream in the fetch spec, but I don't want it to block streaming upload - it is an existing problem for Response.body, too.

Comment (annevk) [reply docs/index.bs:3025]: It became a problem when we added streams. I think now would be a good time to solve it.

And what I mean is that if body is a stream with a null source, service workers needs to transfer ownership back to fetch, such that invoking `req.text()` will start failing afterwards. Presumably this applies to all cases where we return null.

Comment (yutakahirano) [reply docs/index.bs:3025]: Ah, thank you. What do you think about the latest one then?

File: docs/index.bs
Comment (annevk) [inline docs/index.bs:2969]: How is this case not impacted? Is that because there's not even a fetch event for these? Maybe we should return null and null there for clarity to indicate nothing could have changed?

File: docs/index.bs
Comment (annevk) [inline docs/index.bs:2986]: I don't see why we cannot clone here. Body will still have its source afterwards.

File: docs/index.bs
Comment (annevk) [inline docs/index.bs:3044]: You want =Body/unusable= here.

File: docs/index.bs
Comment (annevk) [inline docs/index.bs:3043]: This is unfortunate, but indeed something we're stuck with. Not sure if that's worth pointing out somehow.

**Review by annevk — COMMENTED** [review]
I think this mostly looks good and I think I also misunderstood some things initially. Left suggestions for improvements.

**Comment by yutakahirano** [comment]
Fixed the PR per the discussion at whatwg/fetch#1144.

File: docs/index.bs
Comment (annevk) [inline docs/index.bs:3035]: The second check here is redundant with checking whether it is unusable. In order for a body to be unusable it has to be non-null.

Comment (yutakahirano) [reply docs/index.bs]: Makes sense, but I'd keep this because the next change added a cancel step.

**Review by annevk — APPROVED** [review]
Probably also needs review from a service worker maintainer.

**Comment by yutakahirano** [comment]
@jakearchibald , can you take a look? This is part of whatwg/fetch#1144.

File: docs/index.bs
Comment (MattiasBuelens) [inline docs/index.bs:3037]: The steps are numbered incorrectly, see screenshot:
image

I think it's because the indentation is inconsistent (tabs mixed with spaces). Also, all numbered steps should be prefixed with `1.`, we shouldn't have a `2.` in there.
```suggestion
              1. If |response| is null, |request|'s [=request/body=] is not null, and |request|'s [=request/body=]'s [=Body/source=] is null, then:
                  1. If |request|'s [=request/body=] is [=Body/unusable=], set |handleFetchFailed| to true.
                  1. Else, [=ReadableStream/cancel=] |request|'s [=request/body=] with undefined.
```

**Review by MattiasBuelens — CHANGES_REQUESTED** [review]

File: docs/index.bs
Comment (yutakahirano) [reply docs/index.bs]: Thank you!

**Review by jakearchibald — APPROVED** [review]
LGTM! Glad this turned out so simple.

**Comment by jakearchibald** [comment]
@yutakahirano are you happy for this to be merged?

**Comment by yutakahirano** [comment]
Yeah, thank you!

**Pull request merged**
